app/test/e2e/vocabulary-widget.spec.js

```diff
@@ -0,0 +1,190 @@
+        await requester.post(`/api/v1/dataset/123/widget/${mockWidgetId}/vocabulary/${vocabName}`)
+            .send({ ...vocabData, loggedUser: USERS.ADMIN });
+
+        // Mock the request for dataset validation
+        mockWidget(mockWidgetId);
+
+        // Perform DELETE request for deleting the vocabulary-dataset relationship
+        const response = await requester
+            .delete(`/api/v1/dataset/123/widget/${mockWidgetId}/vocabulary/${vocabName}?loggedUser=${JSON.stringify(USERS.ADMIN)}`)
+            .send();
+
+        // Assert the response as 200 OK with the created data in the body of the request
+        response.status.should.equal(200);
+        response.body.should.have.property('data').and.be.an('array');
+    });
+
+    it('Getting vocab-widget relationships without auth returns 401 Unauthorized', async () => {
+        // Perform GET request for the vocab-widget relationships
+        const response = await requester.post(`/api/v1/dataset/123/widget/123/vocabulary`).send();
+
+        // Assert the response as 401 Unauthorized
+        response.status.should.equal(401);
+        response.body.should.have.property('errors').and.be.an('array');
+        response.body.errors[0].should.have.property('detail').and.equal('Unauthorized');
+    });
+
+    it('Getting vocab-widget relationships with auth returns 200 OK with the requested data', async () => {
+        // Mock the request for widget validation
+        const mockWidgetId = mockWidget().id;
+
+        // Prepare vocabulary test data
+        const vocabName = 'science-widget';
+        const vocabData = { application: 'rw', tags: ['biology', 'chemistry'] };
+
+        // Perform POST request for creating the vocabulary-widget relationship
+        await requester.post(`/api/v1/dataset/123/widget/${mockWidgetId}/vocabulary/${vocabName}`)
+            .send({ ...vocabData, loggedUser: USERS.ADMIN });
+
+        // Mock again the request for widget validation
+        mockWidget(mockWidgetId);
+
+        // Perform GET request for the vocabulary-widget relationships
+        const response = await requester
+            .post(`/api/v1/dataset/123/widget/${mockWidgetId}/vocabulary`)
+            .send({ loggedUser: USERS.ADMIN });
+
+        // Assert the response as 200 OK
+        response.status.should.equal(200);
+        response.body.should.have.property('data').and.be.an('array');
+        response.body.data[0].attributes.should.have.property('name').and.equal(vocabName);
+        response.body.data[0].attributes.should.have.property('application').and.equal('rw');
+        response.body.data[0].attributes.should.have.property('tags').and.deep.equal(['biology', 'chemistry']);
+    });
+
+    afterEach(() => {
+        if (!nock.isDone()) {
+            throw new Error(`Not all nock interceptors were used: ${nock.pendingMocks()}`);
+        }
+    });
+
+    after(async () => {
+        await Resource.deleteMany().exec();
```

Comment:
Can you move all of these from `before` to `beforeEach` and from `after` to `afterEach`? Initially, I had it like this, but then JS's asynchronous nature came to bite me in the ass, as things were happening in weird orders. Thus I started moving towards cleaning up the model in between tests, to ensure all of them are self contained, and dont depend on each other.

Reply:
Done, had to make some changes in the find-by-ids test suite, but it was an easy change 👍